core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -473,274 +387,137 @@ boolean casNextEntryIndex(int ei, int nextOld, int nextNew) {
-     */
-    static private int getValueLengthFromIntArray(int[] long2IntTranslated) {
-        return long2IntTranslated[BLOCK_ID_LENGTH_ARRAY_INDEX] & VALUE_LENGTH_MASK;
-    }
-
     /********************************************************************************************/
     /*-------------- Methods for managing keys and values of a specific entry' -----------------*/
     // Memory Manager interface is via Slice, although Slices are not key/value containers in the
     // EntrySet. MM allocates/releases Slices
 
+    boolean keyLookup(ThreadContext ctx, int ei) {
+        ctx.entryIndex = ei;
+        return readKey(ctx.key, ei);
+    }
+
     /**
-     * buildLookUp builds a LookUp from an entry given by entry index "ei",
+     * Builds a LookUp from an entry given by entry index "ei",
      * so this entry can be referred to later
-     *
-     * If the entry itself or off-heap value is deleted the returned LookUp has null as a Slice
      */
-    LookUp buildLookUp(int ei) {
-        int[] version = new int[1]; // the array of size 1 in order to get output inside it
-        // Atomic snapshot of version and value reference
-        long valueReference = getValueReferenceAndVersion(ei, version);
+    void valueLookUp(ThreadContext ctx) {
+        // Updates the lookup's version and value reference atomically
+        boolean isValid = readValue(ctx.value, ctx.entryIndex);
+
+        /*
+         The value's allocation version indicate the status of the value referenced by {@code value.reference}.
+         If {@code value.reference == INVALID_REFERENCE}, then:
+         {@code version <= INVALID_VERSION} if the removal was completed.
+         {@code version > INVALID_VERSION} if the removal was not completed.
+         otherwise:
+         {@code version <= INVALID_VERSION} if the insertion was not completed.
+         {@code version > INVALID_VERSION} if the insertion was completed.
+         */
 
-        Slice valueSlice = buildValueSlice(valueReference, version[0], memoryManager);
-        if (valueSlice == null) {
+        if (!isValid) {
             // There is no value associated with the given key
-            assert valueReference == INVALID_VALUE_REFERENCE;
             // we can be in the middle of insertion or in the middle of removal
             // insert: (1)reference+version=invalid, (2)reference set, (3)version set
             //          middle state valid reference, but invalid version
             // remove: (1)off-heap delete bit, (2)reference invalid, (3)version negative
             //          middle state invalid reference, valid version
-            return new LookUp(
-                null, valueReference, ei, version[0], getKeyReference(ei),
-                (version[0] >= INVALID_VERSION)); // if version is negative no need to finalize delete
+
+            // if version is negative no need to finalize delete
+            ctx.valueState = (ctx.value.getAllocVersion() < Slice.INVALID_VERSION) ?
```

Comment:
(1) Moved to `EntrySet`.
(2) Fixed. Now the version set even if the reference is invalid so that it will be available even if the reference is invalid.